tfhe/src/shortint/server_key/add.rs

```diff
@@ -5,6 +5,157 @@ use crate::shortint::server_key::CheckError::CarryFull;
 use crate::shortint::{CiphertextBase, PBSOrderMarker};
 
 impl ServerKey {
+    /// Compute homomorphically an addition between two ciphertexts encrypting integer values.
+    ///
+    /// This function, like all "default" operations (i.e. not smart, checked or unchecked), will
+    /// check that the input ciphertext carries are empty and clears them if it's not the case. It
+    /// outputs a ciphertext whose carry is always empty.
+    ///
+    /// This means that when using only "default" operations, a given operation (like add for
+    /// example) has always the same performance characteristics from one call to another and
+    /// guarantees correctness by pre-emptively clearing carries of output ciphertexts.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::shortint::gen_keys;
+    /// use tfhe::shortint::parameters::{PARAM_MESSAGE_2_CARRY_2, PARAM_SMALL_MESSAGE_2_CARRY_2};
+    ///
+    /// // Generate the client key and the server key:
+    /// let (cks, sks) = gen_keys(PARAM_MESSAGE_2_CARRY_2);
+    ///
+    /// let msg = 1;
+    ///
+    /// // Encrypt two messages:
+    /// let ct1 = cks.encrypt(msg);
+    /// let ct2 = cks.encrypt(msg);
+    ///
+    /// // Compute homomorphically an addition:
+    /// let ct_res = sks.add(&ct1, &ct2);
+    ///
+    /// // Decrypt:
+    /// let two = cks.decrypt(&ct_res);
+    /// assert_eq!(msg + msg, two);
+    ///
+    /// let (cks, sks) = gen_keys(PARAM_SMALL_MESSAGE_2_CARRY_2);
+    ///
+    /// // Encrypt two messages:
+    /// let ct1 = cks.encrypt_small(msg);
+    /// let ct2 = cks.encrypt_small(msg);
+    ///
+    /// // Compute homomorphically an addition:
+    /// let ct_res = sks.add(&ct1, &ct2);
+    ///
+    /// // Decrypt:
+    /// let two = cks.decrypt(&ct_res);
+    /// assert_eq!(msg + msg, two);
+    /// ```
+    pub fn add<OpOrder: PBSOrderMarker>(
+        &self,
+        ct_left: &CiphertextBase<OpOrder>,
+        ct_right: &CiphertextBase<OpOrder>,
+    ) -> CiphertextBase<OpOrder> {
+        let tmp_lhs: CiphertextBase<OpOrder>;
+        let tmp_rhs: CiphertextBase<OpOrder>;
+        let lhs: &CiphertextBase<OpOrder>;
+        let rhs: &CiphertextBase<OpOrder>;
+
+        if ct_left.carry_is_empty() {
+            lhs = ct_left;
+        } else {
+            tmp_lhs = self.clear_carry(ct_left);
+            lhs = &tmp_lhs;
+        }
+
+        if ct_right.carry_is_empty() {
+            rhs = ct_right;
+        } else {
+            tmp_rhs = self.clear_carry(ct_right);
+            rhs = &tmp_rhs;
+        }
+
+        let mut ct_res = self.unchecked_add(lhs, rhs);
+        self.clear_carry_assign(&mut ct_res);
+        ct_res
```

Comment:
Here we can clone lhs then call add_assign, we don"t have the same thing as smart operations where we want to clean the carries before the operation

Reply:
hmm disagree, we want to be safe if for whatever reason someone used unchecked right before right ?

Reply:
ah no my bad you are right, I started with the add op not add_assign, good point !

Reply:
but do we want to clone if we can avoid it ?

Reply:
We should end up with the same number of clone. In you current case you don't clone, but use unchecked_add which will clone its lhs before calling`unchecked_add_assign` , in my proposition you clone, and then call `add_assign` which in turns call `unchecked_add_assign` which won't clone

Reply:
We might even save a clone by doing so (in the case lhs has non empty carries)

Reply:
you are perfectly right, will update to use the default assign ops in the default one